### PR TITLE
Tempus-567-Group Support

### PR DESCRIPTION
**Pull request by shubhamkumargupta**
Thank you for submitting a contribution to Tempus.

Cards : #610 #611 #612 #613 #615 #617 
PR contains changes related to Group Support. 
Default Groups has been added for System Admin and Tenant Admin.
AbstractControllerTest has been changed to create default group and assign respective users to group.

In order to streamline the review of the contribution please
ensure the following steps have been taken:

### For all changes:
- [x] Is there a Waffle ticket associated with this PR? Is it referenced 
     in the commit message?

- [x] Does your PR title start with Tempus-567 where 567 is the waffle number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [x] Has your PR been rebased against the latest commit within the target branch (typically dev)?

- [ ] Is your initial contribution a single, squashed commit?

### For code changes:
- [x] Have you ensured that the full suite of tests is executed via mvn clean install at the root Tempus folder?
- [ ] Have you written or updated unit tests to verify your changes?
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)? 
- [ ] Have you added documentation for any UI or API related changes to the /docs folder

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

File: application/src/main/java/com/hashmapinc/server/controller/BaseController.java [hunk 1]
```diff
@@ -274,6 +278,23 @@ Customer checkCustomerId(CustomerId customerId) throws TempusException {
         }
     }
 
+    CustomerGroup checkCustomerGroupId(CustomerGroupId customerGroupId) throws TempusException {
+        try {
+            validateId(customerGroupId, "Incorrect customerGroupId " + customerGroupId);
+            SecurityUser authUser = getCurrentUser();
+            if (authUser.getAuthority() == Authority.SYS_ADMIN ||
+                    (authUser.getAuthority() != Authority.TENANT_ADMIN)) {
```
Comment (anuj-akt) [inline application/src/main/java/com/hashmapinc/server/controller/BaseController.java]: Authority check is being done by Spring Security Expression. Here we should check if the logged in users tenant id same as groups tenant id.

File: application/src/it/java/com/hashmapinc/server/controller/BaseCustomerGroupControllerTest.java [hunk 2]
```diff
@@ -0,0 +1,239 @@
...
+        Collections.sort(loadedCustomerGroups, idComparator);
+
+        Assert.assertEquals(customerGroups, loadedCustomerGroups);
+
+        for (CustomerGroup customerGroup : loadedCustomerGroups) {
+            doDelete("/api/customer/group/"+customerGroup.getId().getId().toString())
+                    .andExpect(status().isOk());
+        }
+    }
+
```
Comment (anuj-akt) [inline application/src/it/java/com/hashmapinc/server/controller/BaseCustomerGroupControllerTest.java]: Refactor these two tests to simplify it. There shouldn't be need of creating so many groups, it will slow down tests.

File: application/src/main/java/com/hashmapinc/server/controller/CustomerGroupController.java [hunk 3]
```diff
@@ -0,0 +1,264 @@
...
+            @PathVariable(CUSTOMER_GROUP_ID) String strCustomerGroupId,
+            @RequestParam int limit,
+            @RequestParam(required = false) String textSearch,
+            @RequestParam(required = false) String idOffset,
+            @RequestParam(required = false) String textOffset) throws TempusException {
+        checkParameter(CUSTOMER_GROUP_ID, strCustomerGroupId);
+        try {
+            CustomerGroupId customerGroupId = new CustomerGroupId(toUUID(strCustomerGroupId));
+            SecurityUser authUser = getCurrentUser();
+            if (authUser.getAuthority() == Authority.CUSTOMER_USER && !authUser.getId().equals(customerGroupId)) {
```
Comment (anuj-akt) [inline application/src/main/java/com/hashmapinc/server/controller/CustomerGroupController.java]: This check is meaningless. Spring Security will only allow tenant admins to access this method.

File: application/src/main/java/com/hashmapinc/server/controller/UserController.java [hunk 4]
```diff
@@ -287,5 +292,66 @@ public void deleteUser(@PathVariable(USER_ID) String strUserId) throws TempusExc
...
+            @PathVariable(USER_ID) String strUserId,
+            @RequestParam int limit,
+            @RequestParam(required = false) String textSearch,
+            @RequestParam(required = false) String idOffset,
+            @RequestParam(required = false) String textOffset) throws TempusException {
+        checkParameter(USER_ID, strUserId);
+        try {
+            UserId userId = new UserId(toUUID(strUserId));
+            SecurityUser authUser = getCurrentUser();
+            if (authUser.getAuthority() == Authority.CUSTOMER_USER && !authUser.getId().equals(userId)) {
```
Comment (anuj-akt) [inline application/src/main/java/com/hashmapinc/server/controller/UserController.java]: This check is meaning less. Either include CUSTOMER_USER role in security expression or remove this check.

File: dao/src/main/java/com/hashmapinc/server/dao/model/nosql/UserEntity.java [hunk 5]
```diff
@@ -21,6 +21,9 @@
 import com.datastax.driver.mapping.annotations.PartitionKey;
 import com.datastax.driver.mapping.annotations.Table;
 import com.fasterxml.jackson.databind.JsonNode;
+import com.hashmapinc.server.common.data.CustomerGroup;
+import com.hashmapinc.server.common.data.id.CustomerGroupId;
```
Comment (anuj-akt) [inline dao/src/main/java/com/hashmapinc/server/dao/model/nosql/UserEntity.java]: Remove unused imports.

File: dao/src/main/java/com/hashmapinc/server/dao/model/sql/UserEntity.java [hunk 6]
```diff
@@ -18,8 +18,10 @@
 
 import com.datastax.driver.core.utils.UUIDs;
 import com.fasterxml.jackson.databind.JsonNode;
+import com.hashmapinc.server.common.data.CustomerGroup;
 import com.hashmapinc.server.common.data.UUIDConverter;
 import com.hashmapinc.server.common.data.User;
+import com.hashmapinc.server.common.data.id.CustomerGroupId;
```
Comment (anuj-akt) [inline dao/src/main/java/com/hashmapinc/server/dao/model/sql/UserEntity.java]: Remove unused imports.

File: dao/src/main/java/com/hashmapinc/server/dao/sql/user/JpaUserDao.java [hunk 7]
```diff
@@ -34,6 +34,8 @@
 import java.util.List;
```
Comment (anuj-akt) [inline dao/src/main/java/com/hashmapinc/server/dao/sql/user/JpaUserDao.java]: This Class is not used anymore. Please revert the changes, it is misleading.

File: dao/src/main/java/com/hashmapinc/server/dao/sql/user/UserRepository.java [hunk 8]
```diff
@@ -16,6 +16,7 @@
  */
```
Comment (anuj-akt) [inline dao/src/main/java/com/hashmapinc/server/dao/sql/user/UserRepository.java]: This Class is not used anymore. Please revert the changes, it is misleading.

File: dao/src/main/java/com/hashmapinc/server/dao/user/CassandraUserDao.java [hunk 9]
```diff
@@ -86,4 +86,9 @@ public User findByEmail(String email) {
         return DaoUtil.convertDataList(userEntities);
     }
 
+    @Override
```
Comment (anuj-akt) [inline dao/src/main/java/com/hashmapinc/server/dao/user/CassandraUserDao.java]: This Class is not used anymore. Please revert the changes, it is misleading.

File: dao/src/main/java/com/hashmapinc/server/dao/user/UserDao.java [hunk 10]
```diff
@@ -59,5 +59,13 @@
      * @return the list of user entities
      */
     List<User> findCustomerUsers(UUID tenantId, UUID customerId, TextPageLink pageLink);
+
```
Comment (anuj-akt) [inline dao/src/main/java/com/hashmapinc/server/dao/user/UserDao.java]: This Class is not used anymore. Please revert the changes, it is misleading.

File: dao/src/main/java/com/hashmapinc/server/dao/user/UserService.java [hunk 11]
```diff
@@ -18,17 +18,22 @@
...
 import com.hashmapinc.server.common.data.page.TextPageLink;
 import com.hashmapinc.server.common.data.security.UserCredentials;
 
+import java.util.List;
+
 public interface UserService {
 	
 	User findUserById(UserId userId);
 
+	TextPageData<User> findUsersByIds(List<UserId> userIds, TextPageLink pageLink);
```
Comment (anuj-akt) [inline dao/src/main/java/com/hashmapinc/server/dao/user/UserService.java]: This Class is not used anymore. Please revert the changes, it is misleading.

File: ui/src/app/api/usergroup.service.js [hunk 12]
```diff
@@ -0,0 +1,178 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+export default angular.module('tempus.api.usergroup', [])
+    .factory('usergroupService', UsergroupService)
+    .name;
+
+/*@ngInject*/
+function UsergroupService($q, $http) {
```
Comment (jayeshkapadnis) [inline ui/src/app/api/usergroup.service.js]: Name should in camel case

File: ui/src/app/api/usergroup.service.js [hunk 13]
```diff
@@ -0,0 +1,178 @@
...
+        $http.post(url, groupIds).then(function success(response) {
+            deferred.resolve(response.data);
+        }, function fail() {
+            deferred.reject();
+        });
+        return deferred.promise;
+
+    }
+
+    function unassignUserToGroup(groupId, userIds) {
```
Comment (jayeshkapadnis) [inline ui/src/app/api/usergroup.service.js]: This should be unassignUserFromGroup similarly for unassignGroupFromUser

File: ui/src/app/components/grid.directive.js [hunk 14]
```diff
@@ -449,14 +453,29 @@ function GridController($scope, $rootScope, $state, $mdDialog, $document, $q, $m
...
 
-        vm.addItemAction = vm.config.addItemAction || {
-                onAction: function ($event) {
-                    addItem($event);
-                },
-                name: function() { return $translate.instant('action.add') },
-                details: function() { return vm.addItemText() },
-                icon: "add"
-            };
+        if (angular.isDefined(vm.config.entType) && vm.config.entType == "usergroup") {
```
Comment (jayeshkapadnis) [inline ui/src/app/components/grid.directive.js]: If you want your own AddItemAction, you need to create Angular controller and attach it to attribute addItemController in grid config of parent controller.  If you just need to change a function from a controller add Item action in list

**Comment by anuj-akt** [comment]
Git log has lots of noise, there are multiple entries for same comment. Need to clean this up.
